src/live_restore/live_restore_fs.c

```diff
@@ -611,30 +634,45 @@ __live_restore_can_service_read(
               "CANNOT SERVICE %s: Reading from hole. Read: %" PRId64 "-%" PRId64 ", hole: %" PRId64
               "-%" PRId64,
               lr_fh->iface.name, offset, read_end, hole->off, WT_EXTENT_END(hole));
-            return (false);
-        } else if (read_begins_in_hole != read_ends_in_hole) {
+            return (NONE);
+        } else if (!read_begins_in_hole && read_ends_in_hole) {
             /*
-             * The read starts in a hole but doesn't finish in it, or vice versa. This breaks
-             * assumptions we make about how the block manager works and is intentionally
-             * unimplemented.
+             * The block manager reads entire pages so we can expect all reads to exist entirely
+             * inside or outside a hole during normal WiredTiger operation. The one exception is
+             * when background migration threads are running as they will copy data chunks
+             * regardless of page size. The background threads always migrate a file from start to
+             * finish so the one case where we partially read from a hole is when the background
+             * thread reads the first part of a page and then we read that page before the second
+             * part is migrated.
              */
-            WT_ASSERT_ALWAYS(session, false, "Read partially covers a hole");
+            __wt_verbose_debug3(session, WT_VERB_FILEOPS,
+              "PARTIAL READ %s: Reading from hole. Read: %" PRId64 "-%" PRId64 ", hole: %" PRId64
+              "-%" PRId64,
+              lr_fh->iface.name, offset, read_end, hole->off, WT_EXTENT_END(hole));
+            *holep = hole;
+            return (PARTIAL);
+        } else if (read_begins_in_hole && !read_ends_in_hole) {
+            /* A partial read should never begin in a hole. */
+            WT_ASSERT_ALWAYS(session, false,
+              "Read (offset: %ld, len: %lu) begins in a hole but does not end in one (offset: %ld, "
+              "len: %lu)",
+              offset, len, hole->off, hole->len);
         }
 
         hole = hole->next;
     }
 
     __wt_verbose_debug3(
       session, WT_VERB_FILEOPS, "CAN SERVICE %s: No hole found", lr_fh->iface.name);
-    return (true);
+    return (FULL);
```

Comment:
Having two paths return `FULL` is a bit confusing. I suggest considering refactoring the code so both `FULL` paths end up here. Maybe either via a `goto done` or a conditional. Also I think getting to this path deserves a comment. How would we get here with `FULL` without hitting the `complete` condition at the beginning since we're holding the extent list lock? My initial thought would be that this thread holds the extent list lock and sees it is not complete. So it would remain incomplete while it holds the lock. In any case, a comment here would help.

Reply:
Getting here without hitting the full condition indicates that a hole list exists for that file but the read is not within any of the holes and therefore the `FULL` read is serviced from the destination.

Reply:
12390ade1